### DISCUSSION TITLE
How to get this array using APITools?

### DISCUSSION THREAD

**Discussion by ghost** [objeck/objeck-lang#162] [General]
`wchar_t const * const * aFilterPatterns`

`APITools_GetCharArray` seems to be not what we need here.

**Comment by objeck** [comment]
I normally use ``APITools_GetStringValue(..)`` and ``APITools_SetStringValue(..)`` which uses a ``wstring``. I will look at the other API today.

**Reply by ghost** [reply]
Sorry my bad. I made such a bad question. Too few information.

`aFilterPatterns` is an argument of a C function on the C library I wanted to use.

On Objeck code, I will map `aFilterPatterns` to an array of Strings. It's either an empty array or contains two Strings "foo" and "bar".

What I'm asking is on the C++ glue code how could I get and extract that array of Strings from Objeck to reconstruct `aFilterPatterns` to feed to the C function. Sorry again if it confused you.

**Reply by ghost** [reply]
I have my solution now. The key is don't map the API 1:1 from C to Objeck. C is a primitive language and Objeck is an advanced language. Why do we have to deal with C's limitations? I mapped `aFilterPatterns` to just a String on Objeck code. The String can be empty or contains "foo" and "bar" with `;` used as the delimiter so I could get and extract that String to reconstruct `aFilterPatterns` on the glue C++ code with just `APITools_GetStringValue`!

**Reply by ghost** [reply]
Your solution is hackish imo. There should be a way to get and extract array of strings from Objeck. A delimited string string just sounds unnatural to me. It should be an array of strings.

**Comment by objeck** [comment]
I think your point is well taken. 

The API tools were not written for general use. We will need to extend the framework to support GTK 3/4. I will refractor and document the framework. The existing framework was created with point solutions rather than holistically designed.

**Comment by objeck** [comment]
Looking into this, is the ask for an API to bidirectionally map Objeck ``Char[,]`` to ``wchar_t**`` and/or ``char**``?

**Reply by objeck** [reply]
Most C libraries are still incompatible with Unicode (i.e., wide characters).

**Reply by ghost** [reply]
It's about getting Objeck's `array of strings` aka `String[,]` on C++ glue code. The library I mentioned is:

https://sourceforge.net/projects/tinyfiledialogs/

It's a small library I use to learn making binding for Objeck. It supports both UTF-8 and UTF-16. The API is ugly and C centric though. I decided to give up on it and find something else that has a nicer API, maybe dwindows.

**Comment by objeck** [comment]
@tn1997tn, please refer to the following [Objeck](https://github.com/objeck/objeck-lang/blob/master/core/lib/experimental/gtk/poc/app_test.obs#L17) and [C++](https://github.com/objeck/objeck-lang/blob/master/core/lib/experimental/gtk/poc/gtk3_test.cpp#L173) PoC for GTK
